File: AwesomeArchComponent/app/src/main/java/com/akexorcist/repositoryarchcomponents/ui/hero/HeroViewModel.java
package com.akexorcist.repositoryarchcomponents.ui.hero;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;

import com.akexorcist.repositoryarchcomponents.api.Resource;
import com.akexorcist.repositoryarchcomponents.api.hero.respond.HeroResult;
import com.akexorcist.repositoryarchcomponents.repo.hero.HeroRepository;
import com.akexorcist.repositoryarchcomponents.util.AbsentLiveData;

/**
 * Created by Akexorcist on 10/6/2017 AD.
 */

public class HeroViewModel extends ViewModel {
    private final HeroRepository heroRepository;
    private LiveData<Resource<HeroResult>> heroResult;
    private final MutableLiveData<Boolean> forceFetchHero = new MutableLiveData<>();

    public HeroViewModel() {
        this.heroRepository = HeroRepository.Companion.getInstance();
        this.heroResult = Transformations.switchMap(forceFetchHero, forceFetch -> {
            if (forceFetch == null) {
                return AbsentLiveData.create();
            }

            return heroRepository.getHero(forceFetch);
        });
    }

    public void getHeroes(boolean forceFetch) {
        if (forceFetch) {
            forceFetchHero.setValue(null);
            heroRepository.clearAllHeroesInDatabase();
        }
        if (forceFetchHero.getValue() != null &&
                isShouldBeSkip(forceFetchHero.getValue(), forceFetch)) {
            forceFetchHero.setValue(forceFetch);
        }
    }

    public LiveData<Resource<HeroResult>> getHeroesResult() {
        return heroResult;
    }

    private boolean isShouldBeSkip(boolean newForceFetch, boolean oldForceFetch) {
        return equals(newForceFetch, oldForceFetch) || (newForceFetch && !oldForceFetch);
    }

    public static boolean equals(Object a, Object b) {
        return (a == b) || (a != null && a.equals(b));
    }
}

File: AwesomeArchComponent/app/src/main/java/com/akexorcist/repositoryarchcomponents/ui/hero/HeroViewModel.kt
package com.akexorcist.repositoryarchcomponents.ui.hero

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel

import com.akexorcist.repositoryarchcomponents.api.Resource
import com.akexorcist.repositoryarchcomponents.api.hero.respond.HeroResult
import com.akexorcist.repositoryarchcomponents.repo.hero.HeroRepository
import com.akexorcist.repositoryarchcomponents.util.AbsentLiveData

/**
 * Created by Akexorcist on 10/6/2017 AD.
 */

class HeroViewModel : ViewModel() {
    private val heroRepository: HeroRepository
    val heroesResult: LiveData<Resource<HeroResult>>
    private val forceFetchHero = MutableLiveData<Boolean>()

    init {
        this.heroRepository = HeroRepository.instance
        this.heroesResult = Transformations.switchMap(forceFetchHero, { forceFetch ->
            if (forceFetch == null) {
                AbsentLiveData.create()
            }
            else
                heroRepository.getHero(forceFetch)
        })
    }

    fun getHeroes(forceFetch: Boolean) {
        if (forceFetch) {
            forceFetchHero.value = null
            heroRepository.clearAllHeroesInDatabase()
        }
        if (forceFetchHero.value != null && isShouldBeSkip(forceFetchHero.value!!, forceFetch)) {
            forceFetchHero.value = forceFetch
        }
    }

    private fun isShouldBeSkip(newForceFetch: Boolean, oldForceFetch: Boolean): Boolean {
        return equals(newForceFetch, oldForceFetch) || newForceFetch && !oldForceFetch
    }

    companion object {

        fun equals(a: Any?, b: Any): Boolean {
            return a === b || a != null && a == b
        }
    }
}
